lib/auth/saml.go

```diff
@@ -347,17 +349,25 @@ func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, e
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	assertionInfo, err := provider.RetrieveAssertionInfo(samlResponse)
+	response, err := provider.ValidateEncodedResponse(samlResponse)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	assertionInfo, err := RetrieveAssertionInfo(provider, connector, response)
 	if err != nil {
 		return nil, trace.AccessDenied(
 			"received response with incorrect or missing attribute statements, please check the identity provider configuration to make sure that mappings for claims/attribute statements are set up correctly. <See: https://goteleport.com/teleport/docs/enterprise/sso/ssh-sso/>, failed to retrieve SAML assertion info from response: %v.", err)
 	}
+	warningInfo, err := provider.VerifyAssertionConditions(&assertionInfo.Assertions[0])
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-	if assertionInfo.WarningInfo.InvalidTime {
+	if warningInfo.InvalidTime {
 		return nil, trace.AccessDenied("invalid time in SAML assertion info")
```

Comment:
nit: if this is a user-facing error - would be helpful to show the invalid value?

Reply:
Not sure, if something goes wrong it's due to clock drift or a delayed response. Would it be helpful to just say that instead?

Reply:
Plus gosaml2 doesn't give us that invalid value, just that it's invalid.

Reply:
I added to the error message that clock drift could be a cause. I don't think I can do anything more.